chi/server.py

```diff
@@ -798,15 +802,34 @@ def associate_floating_ip(server_id, floating_ip_address=None):
         floating_ip_address (str): The IPv4 address of the Floating IP to
             assign. If specified, this Floating IP must already be allocated
             to the project.
+        port_id (str): Optional port ID to assign the floating IP to. If not
+            provided, the will use the first routable port on the server.
 
     """
     if not floating_ip_address:
-        floating_ip_address = get_free_floating_ip()["floating_ip_address"]
-
-    conn = connection()
-    conn.add_ips_to_server(conn.get_server_by_id(server_id), ips=[floating_ip_address])
-
-    return floating_ip_address
+        floating_ip_obj = chi_network.get_free_floating_ip()
+    else:
+        floating_ip_obj = chi_network.get_floating_ip(floating_ip_address)
+
+    conn = connection(session=session())
+    ports = list(conn.network.ports(device_id=server_id))
+    if port_id:
+        port_obj = next(port for port in ports if port["id"] == port_id)
```

Comment:
does this work? I would have assumed that it needs to have brackets to be a list comprehension, e.g. `next([list_comprehension])`

otherwise seems good, thanks!

Reply:
Seems to be the opposite, it fails with brackets since it requires an iterator.